### ISSUE TITLE
Configuration cache deserializes null java.lang.Object when field captured by Groovy closure

### ISSUE THREAD

**Issue by DanielThomas** [gradle/gradle#19992]
We're working our way through out legacy plugins to make them configuration cache compatible. When updating this plugin, the captured `extension` field failed to serialize/deserialize and was null:

https://github.com/nebula-plugins/gradle-info-plugin/pull/97/files#diff-dc352c4922bc4c08233790456f6aca81e9d037f7ca43cfc52266f8efed983eae

Moving to from a field to a local variable fixed it, but appears to be a problem with configuration caching.

### Expected Behavior
Configuration serialization fails and prevents the cache from being reused.

### Current Behavior
The cache is persisted and reused causing the next build to fail.

### Context
Configuration caching should be completely safe to enable without side effects.

### Steps to Reproduce 
I don't have a reduced test case, but you could certainly wind back just this part of the diff if you wanted to reproduce against this project.

### Your Environment
Gradle 7.4

**Comment by nhoughto** [comment]
What does the stacktrace look like for this failure for you? It sounds similar to a problem i'm having serializing Closures: 

```
org.gradle.api.tasks.TaskExecutionException: Execution failed for task ':task'.
	...
Caused by: org.gradle.internal.execution.fingerprint.InputFingerprinter$InputFingerprintingException: Cannot fingerprint input property 'layerFilters': value '[com.bsycorp.gradle.jib.models.LayerFilter@3e15e642]' cannot be serialized.
	...
Caused by: org.gradle.internal.snapshot.ValueSnapshottingException: Could not serialize value of type LayerFilter
	...
Caused by: java.io.NotSerializableException: java.lang.Object
	at org.gradle.internal.snapshot.impl.AbstractValueProcessor.javaSerialization(AbstractValueProcessor.java:171)
	... 74 more
```

Generally it seems like Closures must be able to be serialized via config cache as they are used via `Action<>` everywhere (i think?) but I can't seem to get it to work, the deser seems to blow up when dealing with any instance of a  `Closure` class even if it has been dehydrated (or especially if its been dehydrated and owner etc are `null`). Kinda unrelated but is there a description of best practice of how to capture a lambda/Function/Closure as an `@Input` and not have things explode?

**Comment by DanielThomas** [comment]
This is with Groovy classes seeing null values for captured class-level fields after deserialization:
```
Caused by: org.codehaus.groovy.runtime.typehandling.GroovyCastException: Cannot cast object 'java.lang.Object@6ce3fd36' with class 'java.lang.Object' to class 'nebula.plugin.info.basic.BasicInfoPlugin'
```

Where the object is `null`.

**Comment by nhoughto** [comment]
Hmm ok that is different, losing type info or something after deser. thanks anyway! ill keep digging. 👍

**Comment by DanielThomas** [comment]
Another only caught when reusing the cache:
```
Caused by: groovy.lang.MissingPropertyException: Could not get unknown property 'configurations' for task ':nebula:writeManagedTestDependencies' of type org.gradle.api.DefaultTask.
        at org.gradle.internal.metaobject.AbstractDynamicObject.getMissingProperty(AbstractDynamicObject.java:85)
        at org.gradle.internal.metaobject.AbstractDynamicObject.getProperty(AbstractDynamicObject.java:62)
        at org.gradle.api.DefaultTask_Decorated.getProperty(Unknown Source)
        at org.codehaus.groovy.runtime.InvokerHelper.getProperty(InvokerHelper.java:194)
        at groovy.lang.Closure.getPropertyTryThese(Closure.java:320)
        at groovy.lang.Closure.getPropertyDelegateFirst(Closure.java:310)
        at groovy.lang.Closure.getProperty(Closure.java:296)
        at org.codehaus.groovy.runtime.callsite.PogoGetPropertySite.getProperty(PogoGetPropertySite.java:49)
        at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callGroovyObjectGetProperty(AbstractCallSite.java:341)
        at dependencies_4skn6tmhs5blvdqfx282tc200$_run_closure5$_closure19.doCall(/Users/dannyt/Projects/bb/nebula/nebula/subprojects/core/dependencies.gradle:117)
```

For task:
```
task writeManagedTestDependencies(type: DefaultTask) {
    inputs.files(configurations.managedTestDependencies).withPropertyName("compileClasspath")
    outputs.file(layout.buildDirectory.file("nebula-managed-test-dependencies")).withPropertyName("versionFile")
    doLast {
        def dependencies = configurations.managedTestDependencies.incoming.resolutionResult.allDependencies
        layout.buildDirectory.file("nebula-managed-test-dependencies").get().asFile.withPrintWriter {writer ->
            dependencies.each {
                writer.println(it.selected.moduleVersion)
            }
        }
    }
}
```

**Comment by eskatos** [comment]
Thanks for the report Daniel.

I could reproduce the problem by using:

```groovy
// build.gradle
configurations {
   managedTestDependencies
}

task writeManagedTestDependencies(type: DefaultTask) {
    inputs.files(configurations.managedTestDependencies).withPropertyName("compileClasspath")
    outputs.file(layout.buildDirectory.file("nebula-managed-test-dependencies")).withPropertyName("versionFile")
    doLast {
        def dependencies = configurations.managedTestDependencies.incoming.resolutionResult.allDependencies
        layout.buildDirectory.file("nebula-managed-test-dependencies").get().asFile.withPrintWriter {writer ->
            dependencies.each {
                writer.println(it.selected.moduleVersion)
            }
        }
    }
}
```

And running `gradle --configuration-cache writeManagedTestDependencies` :

- first run storing the cache is successful, no problem reported
- second run loading the cache is successful, no problem reported

But, if you then delete the produced `build/nebula-managed-test-dependencies` directory, then loading the cache fails executing the task with:

```
Execution failed for task ':writeManagedTestDependencies'.
> Could not get unknown property 'configurations' for task ':writeManagedTestDependencies' of type org.gradle.api.DefaultTask.
```

I would have expected problems to be reported because accessing `project.configurations` at execution time is not supported. Same with `project.layout`. So the root problem here is the lack of problem reporting.

That same reproducer can be fixed the following way using Gradle 7.4, if I'm not mistaken about the gathered data:
```groovy
task writeManagedTestDependencies(type: DefaultTask) {
    def resolvedComponents = configurations.managedTestDependencies.incoming.artifacts.resolvedArtifacts.map { it.collect { it.id } }
    inputs.property("resolvedComponents", resolvedComponents)
    outputs.file(layout.buildDirectory.file("nebula-managed-test-dependencies")).withPropertyName("versionFile")
    def layout = project.layout
    doLast {
        layout.buildDirectory.file("nebula-managed-test-dependencies").get().asFile.withPrintWriter {writer ->
            resolvedComponents.get().each {
                writer.println(it.componentIdentifier)
            }
        }
    }
}
```

**Comment by eskatos** [comment]
https://github.com/gradle/gradle/issues/20126 is about properly reporting such problems. Closing this issue as a duplicate
